File: java/leetcode/src/Task33.java
import org.junit.jupiter.api.Test;

public class Task33 {
    public int search(int[] nums, int target) {
        if (nums.length == 0) {
            return -1;
        }
        if (nums.length == 1) {
            return nums[0] == target ? 0 : -1;
        }
        if(target == nums[0]){
            return 0;
        }

        //search the min index
        int left=0,right = nums.length-1;
        int minIdx = -1;
        while(left <= right){
            int mid = left + (right-left)/2;
            if(nums[mid] > nums[(mid+1)%nums.length]){
                minIdx = (mid+1)%nums.length;
                break;
            }
            if(nums[mid] < nums[0]){
                right = mid - 1;
            }else {
                left = mid + 1;
            }
        }

        // binary search
        if(target >  nums[0]){
            left = 0;
            right = (minIdx-1+nums.length)%nums.length;
        }else{
            left = minIdx;
            right = nums.length-1;
        }

        while(left <= right){
            int mid = left + (right-left)/2;
            if(target == nums[mid]){
                return mid;
            }else if(target > nums[mid]){
                left = mid+1;
            }else{
                right = mid-1;
            }
        }

        return -1;
    }


    @Test
    void test() {
        int[] nums = new int[]{8,9,2,3,4};
        int target = 9;
        System.out.println(search(nums, target));
    }
}

File: py/Task33.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        if len(nums) == 1:
            return 0 if nums[0] == target else -1
        # [4,5,6,7,0,1,2]
        # find min idx
        shift = 0
        left, right = 0, len(nums)-1
        while left <= right:
            mid = left + (right-left) // 2
            if nums[mid] < nums[mid-1]:
                shift = mid
                break
            elif nums[mid] < nums[0]:
                right = mid - 1
            else:
                left = mid + 1
        # binary search
        left, right = 0, len(nums)-1
        while left <= right:
            mid = left + (right-left) // 2
            mm = (mid+shift) % len(nums)
            if nums[mm] == target:
                return mm
            elif nums[mm] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1
    # 2020-04-27

    def search2(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        if len(nums) == 1:
            return 0 if nums[0] == target else -1

        shift = 0
        # [4,5,6,7,0,1,2]
        if nums[0] > nums[-1]:
            # binary search, find the min idx
            left, right = 0, len(nums) - 2
            while left <= right:
                mid = (left + right) >> 1
                if nums[mid] > nums[mid+1]:
                    shift = mid + 1
                    break
                elif nums[mid] < nums[left]:
                    right = mid - 1
                else:
                    left = mid + 1

        # find target
        left, right = 0, len(nums)-1
        while left <= right:
            mid = (left + right) >> 1
            mm = (mid+shift) % len(nums)
            if nums[mm] == target:
                return mm
            elif nums[mm] < target:
                left = mid + 1
            else:
                right = mid - 1

        return -1


so = Solution()
nums = [4, 5, 6, 7, 0, 1, 2]
target = 0
print(so.search2(nums, target))
